woocommerce/woocommerce-ios
# Jetpack CP: display benefits banner under a feature flag

Part of #5362 

- [x] ⚠️ Make sure https://github.com/woocommerce/woocommerce-ios/pull/5379 is merged before merging ⚠️ 

## Why

We want to inform users with Jetpack Connection Package (JCP) sites (without Jetpack-the-plugin) about the benefits of installing Jetpack-the-plugin. In the My Store tab, we show a bottom banner for JCP sites with two actions:
- Tapping on the banner shows a screen with Jetpack benefits with a CTA to install Jetpack
- Tapping on the dismiss button dismisses the banner for the site

This PR just contains the logic to display the banner at the bottom of the dashboard tab.

## Changes

Some SwiftUI challenges I ran into:
- The content height is higher than the actual banner height, on launch and after rotation
- After solving the issue above, the content height is 44px higher than the actual banner height when the offline banner is present

I spent quite some time figuring out how to resolve the two issues above. Even though the banner is set with `fixedSize(horizontal: false, vertical: true)` to fix its ideal height to content height, the height is still higher than expected. When the parent view frame changes in UIKit, it does not seem to notify SwiftUI to recalculate the height. In this PR, the solution I ended up with is to set a height constraint and calculate/update the height in `viewDidLayoutSubviews`, and pin the bottom edge with the safe area layout guide to accommodate the offline banner when it is present. Please feel free to suggest other ways to support self-sizing SwiftUI views in UIKit!

## Notes

- Because we use Auto Layout to pin the bottom of the content view, I created two separate constraints for the cases when the content bottom is pinned to the container (no banner) and to the top of the banner. It feels a bit ugly 😞 
- I plan to refactor the banner showing/hiding logic later to use an observable (combining dashboard content view publisher and a boolean publisher to show/hide the banner), when the criteria become async after we persist the banner dismiss state

## Testing

- Launch from PR branch in logged-in state --> there should be a banner shown at the bottom of the My Store tab
- Try rotating the simulator/device, or changing the font sizes --> the banner should stay at the bottom
- (Optional) Try turning the device offline (or rebuilding the app after setting [this line](https://github.com/woocommerce/woocommerce-ios/blob/ff70064acbd6f22da16bd11a8b6013aa0e0f0ded/WooCommerce/Classes/System/WooNavigationController.swift#L112)'s `status` parameter to `notReachable`) --> the banner should stay at the bottom above the offline banner
- Tap anywhere in the banner that is outside of the dismiss CTA --> a detail screen should be shown with all the Jetpack benefits
- Tap "Not Now" or drag down to dismiss the detail screen
- Tap the dismiss CTA ("X") on the banner --> the banner should disappear

## Example screenshots

portrait with offline banner | landscape
-- | --
![Simulator Screen Shot - iPhone 11 - 2021-11-09 at 15 57 42](https://user-images.githubusercontent.com/1945542/140884643-5be0301f-9d78-4b74-b4c6-a31d6cf6665d.png) | ![Simulator Screen Shot - iPhone 11 - 2021-11-09 at 15 35 45](https://user-images.githubusercontent.com/1945542/140884652-8faad612-131c-4eb4-9674-0bb1a9b626d8.png)


Update release notes:
